// wepy-web/lib/app.js
'use strict';

exports.__esModule = true;

var _native = require('./native');

var _native2 = _interopRequireDefault(_native);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _class = function () {
    function _class() {
        _classCallCheck(this, _class);

        this.$addons = {};
        this.$interceptors = {};
        this.$pages = {};
    }

    _class.prototype.$init = function $init(wepy) {
        this.initAPI(wepy);
        this.$wxapp = getApp();
    };

    _class.prototype.use = function use(addon) {
        for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            args[_key - 1] = arguments[_key];
        }

        if (typeof addon === 'string' && this[addon]) {
            this.$addons[addon] = 1;
            this[addon](args);
        } else {
            this.$addons[addon.name] = new addon(args);
        }
    };

    _class.prototype.intercept = function intercept(api, provider) {
        this.$interceptors[api] = provider;
    };

    _class.prototype.promisify = function promisify() {};

    _class.prototype.requestfix = function requestfix() {};

    _class.prototype.initAPI = function initAPI(wepy) {
        var self = this;
        var noPromiseMethods = {
            stopRecord: true,
            pauseVoice: true,
            stopVoice: true,
            pauseBackgroundAudio: true,
            stopBackgroundAudio: true,
            showNavigationBarLoading: true,
            hideNavigationBarLoading: true,
            createAnimation: true,
            createContext: true,
            createCanvasContext: true,
            hideKeyboard: true,
            stopPullDownRefresh: true
        };

        Object.getOwnPropertyNames(wx).forEach(function (key) {
            if (!noPromiseMethods[key] && key.substr(0, 2) !== 'on' && !/\w+Sync$/.test(key)) {
                Object.defineProperty(_native2.default, key, {
                    get: function get() {
                        return function (obj) {
                            obj = obj || {};
                            if (self.$interceptors[key] && self.$interceptors[key].config) {
                                var rst = self.$interceptors[key].config.call(self, obj);
                                if (rst === false) {
                                    if (self.$addons.promisify) {
                                        return Promise.reject('aborted by interceptor');
                                    } else {
                                        obj.fail && obj.fail('aborted by interceptor');
                                        return;
                                    }
                                }
                                obj = rst;
                            }
                            if (key === 'request') {
                                obj = typeof obj === 'string' ? { url: obj } : obj;
                            }
                            if (self.$addons.promisify) {
                                return new Promise(function (resolve, reject) {
                                    var bak = {};
                                    ['fail', 'success', 'complete'].forEach(function (k) {
                                        bak[k] = obj[k];
                                        obj[k] = function (res) {
                                            if (self.$interceptors[key] && self.$interceptors[key][k]) {
                                                res = self.$interceptors[key][k].call(self, res);
                                            }
                                            if (k === 'success') resolve(res);else if (k === 'fail') reject(res);
                                        };
                                    });
                                    wx[key](obj);
                                });
                            } else {
                                var bak = {};
                                ['fail', 'success', 'complete'].forEach(function (k) {
                                    bak[k] = obj[k];
                                    obj[k] = function (res) {
                                        if (self.$interceptors[key] && self.$interceptors[key][k]) {
                                            res = self.$interceptors[key][k].call(self, res);
                                        }
                                        bak[k] && bak[k].call(self, res);
                                    };
                                });
                                wx[key](obj);
                            }
                        };
                    }
                });
                wepy[key] = _native2.default[key];
            } else {
                Object.defineProperty(_native2.default, key, {
                    get: function get() {
                        return function () {
                            for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
                                args[_key2] = arguments[_key2];
                            }

                            return wx[key].apply(wx, args);
                        };
                    }
                });
                wepy[key] = _native2.default[key];
            }
        });
    };

    return _class;
}();

exports.default = _class;
